include/inexor/vulkan-renderer/frame_graph.hpp

```diff
@@ -0,0 +1,419 @@
+    }
+
+    /// @brief Specifies the usage of this texture resource
+    /// @see TextureUsage
+    void set_usage(TextureUsage usage) {
+        m_usage = usage;
+    }
+};
+
+/// @brief A single render stage in the frame graph
+/// @note Not to be confused with a vulkan render pass!
+class RenderStage {
+    friend FrameGraph;
+
+private:
+    const std::string m_name;
+    std::vector<const RenderResource *> m_writes;
+    std::vector<const RenderResource *> m_reads;
+
+    std::vector<VkDescriptorSetLayout> m_descriptor_layouts;
+    std::function<void(const class PhysicalStage *, const wrapper::CommandBuffer &)> m_on_record;
+
+protected:
+    explicit RenderStage(std::string name) : m_name(std::move(name)) {}
+
+public:
+    RenderStage(const RenderStage &) = delete;
+    RenderStage(RenderStage &&) = delete;
+    virtual ~RenderStage() = default;
+
+    RenderStage &operator=(const RenderStage &) = delete;
+    RenderStage &operator=(RenderStage &&) = delete;
+
+    /// @brief Specifies that this stage writes to `resource`
+    void writes_to(const RenderResource &resource);
+
+    /// @brief Specifies that this stage reads from `resource`
+    void reads_from(const RenderResource &resource);
+
+    /// @brief Binds a descriptor set layout to this render stage
+    /// @note This function will soon be removed
+    // TODO(): Refactor descriptor management in the frame graph
+    void add_descriptor_layout(VkDescriptorSetLayout layout) {
+        m_descriptor_layouts.push_back(layout);
+    }
+
+    /// @brief Specifies a function that will be called during command buffer recordation for this stage
+    /// @details This function can be used to specify other vulkan commands during command buffer recordation. The most
+    /// @details common use for this is for draw commands.
+    void set_on_record(std::function<void(const class PhysicalStage *, const wrapper::CommandBuffer &)> on_record) {
+        m_on_record = std::move(on_record);
+    }
+};
+
+class GraphicsStage : public RenderStage {
+    friend FrameGraph;
+
+private:
+    std::unordered_map<const BufferResource *, std::uint32_t> m_buffer_bindings;
+    std::vector<VkPipelineShaderStageCreateInfo> m_shaders;
+
+public:
+    explicit GraphicsStage(std::string &&name) : RenderStage(name) {}
+    GraphicsStage(const GraphicsStage &) = delete;
+    GraphicsStage(GraphicsStage &&) = delete;
+    ~GraphicsStage() override = default;
+
+    GraphicsStage &operator=(const GraphicsStage &) = delete;
+    GraphicsStage &operator=(GraphicsStage &&) = delete;
+
+    /// @brief Specifies that `buffer` should map to `binding` in the shaders of this stage
+    void bind_buffer(const BufferResource &buffer, std::uint32_t binding);
+
+    /// @brief Specifies that `shader` should be used during the pipeline of this stage
+    /// @note Binding two shaders of same type (e.g. two vertex shaders) is undefined behaviour!
+    void uses_shader(const wrapper::Shader &shader);
```

Comment:
This should be added to the TODO list for framegraph validation :)

Reply:
I've created a checklist in #204 :)